Documentation/teaching/labs/block_device_drivers.rst

```diff
@@ -0,0 +1,1160 @@
+but it is not mandatory to finish them, as requests may be finished by other
+parts of the driver.
+
+The :c:data:`lock` parameter, sent when creating a request queue, is a spinlock
+that the kernel holds when executing the request method. For this reason, the
+request function runs in an atomic context and must follow the rules for
+atomic code (it does not need to call functions that can cause sleep, etc.).
+This lock also ensures that no other requests for the device will be added to
+the queue while the request function is running.
+
+Calling the function that processes the request queue is asynchronous relative
+to the actions of any userspace process and no assumptions about the process
+in which the respective function is running should be made. Also, it should not
+be assumed that the buffer provided by a request is from kernel space or user
+space, any operation that accesses the userspace being erroneous.
+
+Below is presented one of the simplest function of type
+:c:type:`request_fn_proc`:
+
+.. code-block:: c
+
+   static void my_block_request(struct request_queue *q)
+   {
+       struct request *rq;
+       struct my_block_dev *dev = q->queuedata;
+
+       while (1) {
+       	   rq = blk_fetch_request(q);
+       	   if (rq == NULL)
+               break;
+
+       	   if (blk_rq_is_passthrough(rq)) {
+               printk (KERN_NOTICE "Skip non-fs request\n");
+               __blk_end_request_all(rq, -EIO);
+              continue;
+       	   }
+
+           /* do work */
+           ...
+
+       	   __blk_end_request_all(rq, 0);
+       }
+   }
+
+The :c:func:`my_block_request` function contains a :c:func:`while` loop for
+iterating through the request queue sent as argument. The operations performed
+within this loop are:
+
+   * Read the first request from the queue using :c:func:`blk_fetch_request`.
+     As described in `Useful functions for processing request queues`_ section,
+     the :c:func:`blk_fetch_request` function retrieves the first item from the
+     request queue and starts the request.
+   * If the function returns NULL, it has reached the end of the request queue
+     (there is no remaining request to be processed) and exits
+     :c:func:`my_block_request`.
+   * A block device can receive calls which do not transfer data blocks (e.g.
+     low level operations on the disk, instructions referring to special ways of
+     accessing the device). Most drivers do not know how to handle these
+     requests and return an error.
+   * To return an error, :c:func:`__blk_end_request_all` function is called,
+     -EIO being the second argument.
+   * The request is processed according to the needs of the associated device.
+   * The request ends. In this case, :c:func:`__blk_end_request_all` function is
+     called in order to complete the request entirely. If all request sectors
+     have been processed, the :c:func:`__blk_end_request` function is used.
+
+.. _bio_structure:
```

Comment:
Please remove this comment, it doesn't have any value.

Reply:
It does have. Please check line 491. because the subsection title contains `` (back quotes), when I reference the subsection I had to use :ref: